File: 02comm/03uart_rw.js
var SerialPort = require('serialport');
var port = new SerialPort("/dev/serial0", {baudRate:9600}, function(err){
	if(err){
		console.log('err:' + err.message);
		port = null;
	}else{
		var line = '';
		port.on('data', function(data){
			if(data =='\r'){
				console.log('recv:' + line +'\r');	
				line = '';
			}else{
				line +=data;
			}
			
		});
	}
});
if(port){
	process.stdout.write('>');
	process.stdin.on('data', function(data){
		port.write(data+'\r');
		process.stdout.write('>');
	});
}

File: 02comm/03uart_rw.py
import serial, threading

port = serial.Serial("/dev/serial0", baudrate=9600, timeout=3.0)

class ReadThread(threading.Thread):
    flag = True
    
    def run(self):
        while self.flag:
            line = self.readline()
            print("recv:" +line)
        print("read thread died.")
        
    
    def readline(self):
        line = ''
        while True:
            ch = port.read().decode()
            if ch =='\r' or ch=='\n' or ch == '':
                if len(line) > 0 :
                    return line
            else:
                line +=ch
    
try:
    if port.isOpen() :
        t = ReadThread()
        t.start()
        print('serial is ready.')
        while True:
            line = input(">")
            if line == 'exit':
                t.flag = False
                break
            print(line)
            port.write( (line + '\r\n').encode())
        print('bye')
finally:
    port.close()